Use webdriver to set permission on getUserMedia calls

dontcallmedom opened the pull request:
this includes removing the default fake-ui from chrome wpt runner - this makes obtaining permission a prerequisite for using `getUserMedia` in tests.

dontcallmedom committed:
Use webdriver to set permission on getUserMedia calls

Rewrite mediacapture-streams using async/await for consistency in the process
Move a couple of manual tests back to automated as a result

dontcallmedom committed:
Remove fake-ui from chrome wptrunner

Rely instead of SetPermission webdriver hook; this allows testing permission denial.

reillyeon reviewed (approved):
Overall this change seems reasonable but I only took a close look at `mediacapture-image/ImageCapture-creation.https.html` since that's the file I'm an owner for.

alvestrand commented on mediacapture-record/MediaRecorder-bitrate.https.html:
```diff
@@ -5,6 +5,9 @@
 <script src="/resources/testharnessreport.js"></script>
+<script src=/resources/testdriver.js></script>
+<script src=/resources/testdriver-vendor.js></script>
+<script src='../mediacapture-streams/permission-helper.js'></script>
```
can we please have one convention on how filenames are quoted in "script src=" statements?
This makes 3 different versions in a single file.

alvestrand commented on mediacapture-streams/MediaStream-MediaElement-preload-none.https.html:
```diff
@@ -0,0 +1,83 @@
+<html>
+    <head>
+        <title>Test that the HTMLMediaElement preload 'none' attribute value is ignored for MediaStream used as srcObject and MediaStream object URLs used as src.</title>
+        <link rel="author" title="Matthew Wolenetz" href="mailto:wolenetz@chromium.org"/>
```
This file seems like it belongs to a different PR.

dontcallmedom replied:
that file was a `-manual` file, presumably because when it was created, one would not gain access to getUserMedia. Now that we do, it feels reasonable to make the move here.

alvestrand commented on mediacapture-streams/permission-helper.js:
```diff
@@ -0,0 +1,10 @@
+    for (let s of scope) {
+      await test_driver.set_permission({ name: s }, status, true);
+    }
+  } catch {
```
generic catch is usually a bad idea. What error do we expect here?

dontcallmedom replied:
this is to deal with browsers that don't implement the `set_permission` webdriver command (i.e. both Firefox and Safari at this time point).

alvestrand replied:
so we should catch "no such function". I'm worried about the case where it fails for an unrelated reason (such as calling with "scope=['cammera']" and getting a "no such value" exception - that could be hard to debug when the error is caught and ignored).

alvestrand commented on mediacapture-streams/permission-helper.js:
```diff
@@ -0,0 +1,10 @@
+async function setPermission(status="granted", scope=["camera", "microphone"]) {
```
the "status" argument is never used. Is it a future goal to be able to set setPermission(status=denied, scope=["camera"])? If so, document (or even better, implement).

dontcallmedom replied:
it's used in `GUM-deny.html`

alvestrand commented:
So the harness-setting-permission function has finally landed? Great!

A couple of things:

- the name setPermission() is a bad idea. setMediaDevicePermission() is longer, but won't conflict with everyone else's helpers for setting their particular permissions. When we go to tests that mix mulitple interfaces with separate permissions they want to set, it's important that they don't get to clobber each other.

- setting permission for camera when microphone only is needed by the test is sloppy and will fail to detect errors involving checking the wrong permission

- this is a murderous lot of file changes for doing the same thing in multiple places. Would it be possible to wrap this inside a helper function rather than repeating the same 3 lines over and over again?

- I can't see that this PR actually removes the Chrome flag. So the tests will pass on Chrome whether or not the test driver function exists or not. Is this intentional, or is it "delayed to the next PR"?
-

dontcallmedom commented:
>     * the name setPermission() is a bad idea. setMediaDevicePermission() is longer, but won't conflict with everyone else's helpers for setting their particular permissions. When we go to tests that mix mulitple interfaces with separate permissions they want to set, it's important that they don't get to clobber each other.

Will fix to `setMediaDevicePermission()`
 
>     * setting permission for camera when microphone only is needed by the test is sloppy and will fail to detect errors involving checking the wrong permission

I can take a pass at setting permissions more specifically.

>     * this is a murderous lot of file changes for doing the same thing in multiple places. Would it be possible to wrap this inside a helper function rather than repeating the same 3 lines over and over again?

I sure wish it was; but doing this cleanly would require using JS modules, which I don't think the WPT infrastructure is ready for at this point.

>     * I can't see that this PR actually removes the Chrome flag. So the tests will pass on Chrome whether or not the test driver function exists or not. Is this intentional, or is it "delayed to the next PR"?

It does remove the flag, see https://github.com/web-platform-tests/wpt/pull/31847/files#diff-0df24b5b583c460182e687f7dc7a6a79dd2cd3389bc4a96f48483f60fceb51f7

dontcallmedom committed:
Catch web driver error more specifically when setting permission

per https://github.com/web-platform-tests/wpt/pull/31847#pullrequestreview-824508906

dontcallmedom committed:
Rename setPermission in setMediaPermission

per https://github.com/web-platform-tests/wpt/pull/31847#issuecomment-987261583

dontcallmedom committed:
Limit media capture permissions to what is strictly needed for the test

per https://github.com/web-platform-tests/wpt/pull/31847#issuecomment-987261583

dontcallmedom committed:
Align quoting styles for script loading

alvestrand reviewed (approved):
This is a great step forward in testing permission-gated functionality!
